token: inline optimizations

bhalevy opened the pull request:
This series contains several optimizations for dht::token
around its comparison functions as well as minimum_token and maximum_token definitions,
by moving them inline into dht/token.hh

This results in a nice improvement in perf-simple-query:
```
==> perf-simple-query.pre <== (21c67a5a649fe0d65323e3f5171d5dced619ef0f)
         throughput: mean=95774.01 standard-deviation=1129.83 median=96243.64 median-absolute-deviation=1090.08 maximum=96864.09 minimum=94471.19
instructions_per_op: mean=41813.68 standard-deviation=16.27 median=41809.29 median-absolute-deviation=7.02 maximum=41841.64 minimum=41799.41
  cpu_cycles_per_op: mean=22383.19 standard-deviation=331.01 median=22254.53 median-absolute-deviation=332.26 maximum=22744.11 minimum=21996.73

==> perf-simple-query.post.0 <== (token: move ordering operator inline)
         throughput: mean=96350.01 standard-deviation=640.10 median=96228.88 median-absolute-deviation=621.45 maximum=96988.16 minimum=95478.51
instructions_per_op: mean=41627.13 standard-deviation=37.55 median=41627.06 median-absolute-deviation=2.43 maximum=41679.44 minimum=41573.31
  cpu_cycles_per_op: mean=22184.65 standard-deviation=151.03 median=22163.05 median-absolute-deviation=120.83 maximum=22348.49 minimum=21967.30

==> perf-simple-query.post.1 <== (token: operator<=>: optimize the common case)
         throughput: mean=96778.29 standard-deviation=1719.34 median=97021.72 median-absolute-deviation=1059.56 maximum=98300.99 minimum=93893.75
instructions_per_op: mean=41590.25 standard-deviation=5.53 median=41589.50 median-absolute-deviation=4.17 maximum=41598.39 minimum=41584.57
  cpu_cycles_per_op: mean=22135.33 standard-deviation=471.98 median=21969.30 median-absolute-deviation=244.89 maximum=22905.24 minimum=21685.33

==> perf-simple-query.post.3 <== (token: always initialize data member)
         throughput: mean=98264.33 standard-deviation=998.49 median=98533.02 median-absolute-deviation=780.45 maximum=99075.40 minimum=96656.51
instructions_per_op: mean=41657.61 standard-deviation=22.53 median=41648.49 median-absolute-deviation=12.89 maximum=41696.81 minimum=41642.07
  cpu_cycles_per_op: mean=21808.57 standard-deviation=93.63 median=21794.56 median-absolute-deviation=75.41 maximum=21949.46 minimum=21719.55

==> perf-simple-query.post.4 <== (token: constexpr ctors, methods, and minimum/maximum_token)
         throughput: mean=98095.05 standard-deviation=1333.32 median=98930.22 median-absolute-deviation=906.80 maximum=99209.38 minimum=96194.25
instructions_per_op: mean=41572.28 standard-deviation=6.04 median=41574.49 median-absolute-deviation=4.76 maximum=41579.56 minimum=41564.72
  cpu_cycles_per_op: mean=21831.35 standard-deviation=169.56 median=21732.86 median-absolute-deviation=102.93 maximum=22091.66 minimum=21689.63

==> perf-simple-query.post.5 <== (token: initialize non-key tokens with min() value)
         throughput: mean=99502.32 standard-deviation=1003.70 median=99744.03 median-absolute-deviation=388.87 maximum=100482.95 minimum=97813.42
instructions_per_op: mean=41593.48 standard-deviation=17.27 median=41585.25 median-absolute-deviation=8.46 maximum=41619.41 minimum=41575.86
  cpu_cycles_per_op: mean=21545.90 standard-deviation=86.66 median=21578.01 median-absolute-deviation=43.17 maximum=21612.41 minimum=21395.42
```

Optimization only. No backport required

raphaelsc commented on dht/token.cc:
```diff
@@ -18,6 +18,14 @@ namespace dht {
 
+inline int64_t long_token(const token& t) {
+    if (t.is_minimum() || t.is_maximum()) {
+        return std::numeric_limits<int64_t>::min();
```
I was going to suggest we could rewrite 't.is_minimum() || t.is_maximum()' as t._kind != kind::key, but compilers are smart enough to do that using bitwise.

```
f(token&):                            # @f(token&)
        test    dword ptr [rdi], -3
        jne     .LBB0_2
        movabs  rax, -9223372036854775808
        ret
.LBB0_2:
        mov     rax, qword ptr [rdi + 8]
        ret
```

-3 means only bit 1 is zeroed, which refers to key kind.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10435/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10435/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10652/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10435/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2633/)
#### Build Details:
- Duration: 3 hr 12 min
- Builder: spider4.cloudius-systems.com

bhalevy commented on dht/token.hh:
```diff
@@ -34,18 +34,20 @@ class token {
     // (So INT64_MAX is twice as likely as every other token.)
-    static inline int64_t normalize(int64_t t) {
+    static constexpr int64_t normalize(int64_t t) {
         return t == std::numeric_limits<int64_t>::min() ? std::numeric_limits<int64_t>::max() : t;
```
We do:
```
bhalevy@lt scylla$ git grep -En '\[\[(un)?likely\]\]' | wc -l
95
```

But we use it sparingly to optimize common paths vs. exceptional paths that are rare by design.
But otherwise, for things that are workload-dependent, we would let branch prediction do its thing.

michoecho replied:
@mykaul Why are you asking about `likely` in this context?

mykaul replied:
"If a token with value INT64_MIN is generated" made me wonder how common is this scenario.

bhalevy replied:
> "If a token with value INT64_MIN is generated" made me wonder how common is this scenario.

With vnodes it would be due to hashing and so very rare.
With tablets, there is no such chance since we prepare the tablets token ranges starting from `dht::first_token()` which has the value of `min() + 1`

avikivity commented on dht/token.hh:
```diff
@@ -213,7 +213,25 @@ struct raw_token_less_comparator {
+
+inline std::strong_ordering operator<=>(const token& t1, const token& t2) {
+    if (t1._kind < t2._kind) {
+        return std::strong_ordering::less;
```
We should remove token::_kind one day.

bhalevy replied:
Yeah.  @michoecho started some work in this direction.  Curreny we have too many ways to represent ranges.  I think that before/after all tokens concept belongs to bounds.
And then there are intervals with missing bounds that also stand for infinity (a.k.a unbound/semi-bound ranges)

bhalevy replied:
Maybe we should simply keep tokens as signed 128 bit numbers and then the kind would be essentially in the most significant word.

michoecho replied:
> Yeah. @michoecho started some work in this direction.

More than just started. Removing `token::_kind` was indeed one of the things done by https://github.com/scylladb/scylladb/pull/8806, more than 3 years ago.

That PR was basically finished, but at the time I was too much of a noob and too scared of breaking something (especially in the thrift API) to get it merged.

bhalevy replied:
> Maybe we should simply keep tokens as signed 128 bit numbers and then the kind would be essentially in the most significant word.

This idea might have merit...
See https://godbolt.org/z/5jWv5bsef

avikivity replied:
> > Yeah. @michoecho started some work in this direction.
> 
> More than just started. Removing `token::_kind` was indeed one of the things done by #8806, more than 3 years ago.
> 
> That PR was basically finished, but at the time I was too much of a noob and too scared of breaking something (especially in the thrift API) to get it merged.

Now that you're experienced you should be even more scared.

I think the hard parts are when crossing node boundaries, since you have to work with old nodes. But I expect all tokens there will be part of intervals, so we can introduce helpers to convert interval\<heavy_token> to interval\<token>.

avikivity commented on dht/token.hh:
```diff
@@ -42,7 +42,9 @@ public:
-    token() : _kind(kind::before_all_keys) {
+    token()
+        : _kind(kind::before_all_keys)
+        , _data(0) {
```
Should we initialize it to min()? then long_token() doesn't need to check the kind.

bhalevy replied:
That should work since we don't rely on a particular value in data for min/max tokens.

bhalevy replied:
@avikivity actually I take my words back, as the existing code compares tokens against the minimum of maximum tokens like the following:
```diff
diff --git a/db/size_estimates_virtual_reader.cc b/db/size_estimates_virtual_reader.cc
index a8979f66d1..b9fd4cecf4 100644
--- a/db/size_estimates_virtual_reader.cc
+++ b/db/size_estimates_virtual_reader.cc
@@ -195,10 +195,10 @@ static future<std::vector<token_range>> get_local_ranges(replica::database& db,
         // All queries will be on that table, where all entries are text and there's no notion of
         // token ranges form the CQL point of view.
         auto left_inf = boost::find_if(ranges, [] (auto&& r) {
-            return r.end() && (!r.start() || r.start()->value() == dht::minimum_token());
+            return r.end() && (!r.start() || r.start()->value().is_minimum());
         });
         auto right_inf = boost::find_if(ranges, [] (auto&& r) {
-            return r.start() && (!r.end() || r.end()->value() == dht::maximum_token());
+            return r.start() && (!r.end() || r.end()->value().is_maximum());
         });
         if (left_inf != right_inf && left_inf != ranges.end() && right_inf != ranges.end()) {
             local_ranges.push_back(token_range{to_bytes(right_inf->start()), to_bytes(left_inf->end())});
```

And currently min/max tokens are using 0 in the _data member.

so if using `min()` for the in-memory token, we need to serialize the data value as 0
and convert it to std::min upon deserialization.
I'm not sure it's worth it...

avikivity commented:
Nice. I left a minor suggestion.

bhalevy commented on dht/token.hh:
```diff
@@ -229,7 +231,9 @@ inline std::strong_ordering operator<=>(const token& t1, const token& t2) {
-inline bool operator==(const token& t1, const token& t2) { return t1 <=> t2 == 0; }
+inline bool operator==(const token& t1, const token& t2) {
+    return t1._data == t2._data && t1._kind == t2._kind;
+}
```
True, but I wanted to compare `data` first since for different keys we don't need to test kind at all.

bhalevy replied:
Need to check the x86 assembly.
It's possible that the default implementation can be optimized (e.g by using vector instructions)

michoecho replied:
The compiler is going to do this branchlessly anyway, so the order doesn't matter.

And even if it did generate a branch, you shouldn't rely on the source code order for generating a particular order in the executable. `[[likely]]` would be probably more appropriate tool for that than shifting the source around.

avikivity replied:
I don't think the compiler will make it branchless.  We can trick it with !((t1._data - t2._data) | (t1._kind - t2._kind)).

Intel APX introduces conditional compare instructions, so in about 20-25 years it will be branchless.

avikivity commented:
Here's an evil trick: encode it as __int128, min_token = 0xffff'ffff'ffff'fffe'8000'0000'0000'0000, regular = sign-extended token, max_token = 0x0000'0000'0000'0001'8000'0000'0000'0000. Then all compares are just arithmetic compares, and long_token is just reading the low-order bits.

However, it doesn't add any performance AFAICT compared to your work (after you initialize _data with min()).

(author is not responsible for nibble counts in literals)

avikivity commented:
Ahhh! relying on initialization of _data doesn't work across node boundaries, since tokens received from old nodes will not have _data initialized. Must sanitize on rpc boundary!

bhalevy commented:
> Here's an evil trick: encode it as __int128, min_token = 0xffff'ffff'ffff'fffe'8000'0000'0000'0000, regular = sign-extended token, max_token = 0x0000'0000'0000'0001'8000'0000'0000'0000. Then all compares are just arithmetic compares, and long_token is just reading the low-order bits.
> 
> However, it doesn't add any performance AFAICT compared to your work (after you initialize _data with min()).
> 
> (author is not responsible for nibble counts in literals)

https://github.com/scylladb/scylladb/pull/19782#discussion_r1683216217

bhalevy commented:
> Ahhh! relying on initialization of _data doesn't work across node boundaries, since tokens received from old nodes will not have _data initialized. Must sanitize on rpc boundary!

The `token(kind k, const bytes& b)` ctor (and same with bytes_view) take care of that, don't they?

avikivity commented:
> > Here's an evil trick: encode it as __int128, min_token = 0xffff'ffff'ffff'fffe'8000'0000'0000'0000, regular = sign-extended token, max_token = 0x0000'0000'0000'0001'8000'0000'0000'0000. Then all compares are just arithmetic compares, and long_token is just reading the low-order bits.
> > However, it doesn't add any performance AFAICT compared to your work (after you initialize _data with min()).
> > (author is not responsible for nibble counts in literals)
> 
> [#19782 (comment)](https://github.com/scylladb/scylladb/pull/19782#discussion_r1683216217)

Aha. But need to be careful about cross-node boundaries.

avikivity commented:
> > Ahhh! relying on initialization of _data doesn't work across node boundaries, since tokens received from old nodes will not have _data initialized. Must sanitize on rpc boundary!
> 
> The `token(kind k, const bytes& b)` ctor (and same with bytes_view) take care of that, don't they?

Maybe, need to check carefully. Maybe we were lucky with the previous variable-size representation.

bhalevy commented:
> > > Ahhh! relying on initialization of _data doesn't work across node boundaries, since tokens received from old nodes will not have _data initialized. Must sanitize on rpc boundary!
> > 
> > 
> > The `token(kind k, const bytes& b)` ctor (and same with bytes_view) take care of that, don't they?
> 
> Maybe, need to check carefully. Maybe we were lucky with the previous variable-size representation.

This is how the current token serializer is implemented, using the token::data() method to get its `bytes` representation:
```c++
void serializer<dht::token>::write(Output& buf, const dht::token& obj) {
  set_size(buf, obj);
  static_assert(is_equivalent<decltype(obj._kind), dht::token::kind>::value, "member value has a wrong type");
  serialize(buf, obj._kind);
  static_assert(is_equivalent<decltype(obj.data()), bytes>::value, "member value has a wrong type");
  serialize(buf, obj.data());
}
```

And for deserializing we construct the token using the `token(kind, bytes)` constructor:
```c++
dht::token serializer<dht::token>::read(Input& buf) {
 return seastar::with_serialized_stream(buf, [] (auto& buf) {
  size_type size = deserialize(buf, boost::type<size_type>());
  auto in = buf.read_substream(size - sizeof(size_type));
  auto __local_1 = deserialize(in, boost::type<dht::token::kind>());
  auto __local_2 = deserialize(in, boost::type<bytes>());

  dht::token res {std::move(__local_1), std::move(__local_2)};
  return res;
 });
}
```

So I think we should be fine.
Need to put this to test of course.

bhalevy commented:
> Here's an evil trick: encode it as __int128, min_token = 0xffff'ffff'ffff'fffe'8000'0000'0000'0000, regular = sign-extended token, max_token = 0x0000'0000'0000'0001'8000'0000'0000'0000. Then all compares are just arithmetic compares, and long_token is just reading the low-order bits.
> 
> However, it doesn't add any performance AFAICT compared to your work (after you initialize _data with min()).
> 
> (author is not responsible for nibble counts in literals)

@avikivity I have a working implementation using int128_t, but it degraded performance a bit.
I thing due to the need to access the kind member, and the need to convert it to the pre-existing enum values.
See https://github.com/bhalevy/scylla/commit/a936d3f726a1be46f891bd67701d5e920e4fddf1

I'm abandoning this direction for now....

bhalevy commented:
In v2 ([365e1fb](https://github.com/scylladb/scylladb/commit/365e1fb1b95bf3183b34ad1120950bf5b5ae3927)):
- partitioner_test: add more token-level tests
- token: operator<=>: optimize the common case
  - Just use `[[likely]]``
- token: always initialize data member
  - define `token::operator==() = default`
- token: constexpr ctors, methods, and minimum/maximum_token
  - cleaned up patch
  - added last_token and is_last for completeness
- token: make kind-based ctor private
  - prepare for next patch, with no functional changes
- token: initialize non-key tokens with min() value
  - based on https://github.com/scylladb/scylladb/pull/19782#discussion_r1683146043
  - This should be safe, upgrade-wise, based on auditing the serialization/deserialization paths.
    - The in-memory value of the `_data` member is decoupled from the serialized value, so we don't care what's sent on the wire, and initialize the in-memory value to 0 up-to this series, and to min() with it. But both sides of the wire ignore the on-wire value.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10480/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10480/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/mvcc_test
:small_blue_diamond: boost/network_topology_strategy_test
:small_blue_diamond: boost/partitioner_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/tablets_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10480/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2668/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10689/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10480/)
#### Build Details:
- Duration: 3 hr 47 min
- Builder: spider2.cloudius-systems.com

avikivity commented:
> > Here's an evil trick: encode it as __int128, min_token = 0xffff'ffff'ffff'fffe'8000'0000'0000'0000, regular = sign-extended token, max_token = 0x0000'0000'0000'0001'8000'0000'0000'0000. Then all compares are just arithmetic compares, and long_token is just reading the low-order bits.
> > However, it doesn't add any performance AFAICT compared to your work (after you initialize _data with min()).
> > (author is not responsible for nibble counts in literals)
> 
> @avikivity I have a working implementation using int128_t, but it degraded performance a bit. I thing due to the need to access the kind member, and the need to convert it to the pre-existing enum values. See [bhalevy@a936d3f](https://github.com/bhalevy/scylla/commit/a936d3f726a1be46f891bd67701d5e920e4fddf1)
> 
> I'm abandoning this direction for now....

That's okay. A better direction (but more difficult) is to remove _kind completely.
